error.go

```diff
@@ -54,6 +55,113 @@ func RootCause(err error) error {
 	}
 }
 
+// errf is a version of fmt.Errorf with support for a chain of multiple
+// formatted error messages.
+//
+// After msg, N arguments are consumed as formatting arguments for that
+// message, where N is the number of % symbols in msg. Following that, another
+// string may be added to become the next error in the chain. Each new error
+// is the `cause()` for the prior error.
+//
+//   err := errf(
+//     "could not process %v", thing,
+//     "name %q is invalid", thing.Name,
+//  )
+//  fmt.Println(err)  // could not process Thing: name Foo is invalid
+//  fmt.Println(RootCause(err))  // name Foo is invalid
+//
+// In place of a string, the last error can be another error, in which case it
+// will be treated as the cause of the prior error chain.
+//
+//   errf(
+//     "could not process %v", thing,
+//     "date %q could not be parsed", thing.Date,
+//     parseError,
+//  )
+func errf(msg string, args ...interface{}) error {
+	// To ensure that errf controls all calls to buildErrf, we make it a
+	// closure instead of its own function. This allows buildErrf to assume
+	// that len(args) > 0 for all calls.
```

Comment:
To me this sentences implies that making buildErrf being in the closure allows for len(arg) > 0 to be assumed. I don't think that is quite true, what I do think is true is that in ensures that buildErrf can only be used through errF, which limits API surface.

Reply:
It's true because errf has at least 1 argument due to it's signature and the way the recursion is written, the base case is buildErrf is called with 1 argument. 

Still, this comment doesn't really clarify much for someone reading this code - better to just explain that buildErrf is a closure being called recursively.

Reply:
Reworded